src/com/ms/silverking/log/Log.java

```diff
@@ -19,428 +18,445 @@
+  }
+
+  public static Level getLevel() {
+    return __level;
+  }
+
+  public static void setLevel(String logLevel) {
+    setLevel(Level.parse(logLevel));
+  }
+
+  public static void setLevel(Level level) {
+    synchronized (Log.class) {
+      _level = level.intValue();
+      __level = level;
+      logDest.setLevel(level);
+    }
+  }
+
+  public static void setLevelAll() {
+    setLevel(Level.ALL);
+  }
+
+  public static void setLevelOff() {
+    setLevel(Level.OFF);
+  }
+
+  public static void setLevelLog() {
+    setLevel(Level.CONFIG);
+  }
+
+  public static void setPrintStreams(String fileName) throws IOException {
+    setPrintStreams(new File(fileName));
+  }
+
+  public static void setPrintStreams(File file) throws IOException {
+    setPrintStreams(new FileOutputStream(file));
+  }
+
+  public static void setPrintStreams(OutputStream out) {
+    logDest.setPrintStreams(out);
+  }
+
+  public static void countdownWarning(String m, int countdownSeconds) {
+    countdown(Level.WARNING, m, countdownSeconds);
+  }
+
+  public static void countdown(Level level, String m, int countdownSeconds) {
+    for (int t = countdownSeconds; t > 0; t--) {
+      Log.log(level, "Countdown: " + m + ". " + t + " seconds remaining.");
+      ThreadUtil.sleepSeconds(1);
+    }
+    Log.log(level, "Countdown: " + m + ". Countdown complete.");
+  }
+
+  public static void log(Level level, String m) {
+    if (alertReceiver != null && level.intValue() >= alertLevel) {
+      alertReceiver.sendAlert(new Alert(alertContext, level.intValue(), alertKey != null ? alertKey : m, m, alertData));
+    }
+    logDest.log(level, m);
+  }
+
+  public static void warning(String m, Object o) {
+    if (levelMet(Level.WARNING)) {
+      warning(m + o);
+    }
+  }
+
+  public static void warning(Object o) {
+    if (o == null) {
+      warning("null");
+    } else {
+      warning(o.toString());
+    }
+  }
+
+  public static void warning(String m) {
+    log(Level.WARNING, m);
+  }
+
+  public static void warningAsyncf(String f, Object... args) {
+    if (levelMet(Level.WARNING)) {
+      warningAsyncInternal(new LogEntry(Level.WARNING, String.format(f, args)));
+    }
+  }
+
+  public static void warningf(String f, Object... args) {
+    if (levelMet(Level.WARNING)) {
+      warning(String.format(f, args));
+    }
+  }
+
+  public static void warningAsync(String m, Object o) {
+    if (levelMet(Level.WARNING)) {
+      warningAsyncInternal(new LogEntry(Level.WARNING, o, m));
+    }
+  }
+
+  public static void warningAsync(Object o) {
+    if (o == null) {
+      warningAsync("null");
+    } else {
+      warningAsync(o.toString());
+    }
+  }
+
+  public static void warningAsync(String m) {
+    if (levelMet(Level.WARNING)) {
+      warningAsyncInternal(new LogEntry(Level.WARNING, m));
+    }
+  }
+
+  private static void warningAsyncInternal(LogEntry le) {
+    try {
+      logQueue.put(le);
+    } catch (InterruptedException ie) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: InterruptedException is ignored. This can delay thread shutdown and clear the thread’s interrupt status. Only code that implements a thread’s interruption policy can swallow an interruption request.

**Fix**: Rethrow the InterruptedException or reinterrupt the current thread using *Thread.currentThread().interrupt()* so that higher-level interrupt handlers can function correctly.
 If you are wrapping the InterruptedException inside a RuntimeException, call *Thread.currentThread().interrupt()* before throwing the RuntimeException.

Learn more about [interrupts](https://docs.oracle.com/javase/tutorial/essential/concurrency/interrupt.html) and [dealing with InterruptedException](https://www.ibm.com/developerworks/java/library/j-jtp05236/index.html?ca=drs-#N10187)